packages/dao-contracts/src/dao-contract.ts

```diff
@@ -0,0 +1,95 @@
+import { Token } from '@bitcoin-computer/TBC20'
+import { Contract } from '@bitcoin-computer/lib'
+
+type ElectionType = {
+  proposalMod: string
+  tokenRoot: string
+  description: string
+}
+
+type VoteType = {
+  electionId: string
+  tokens: Token[]
+  vote: 'accept' | 'reject'
+}
+
+export class Election extends Contract {
+  proposalMod!: string
+  tokenRoot!: string
+  description!: string
+  constructor({ proposalMod, tokenRoot, description }: ElectionType) {
+    super({ proposalMod, tokenRoot, description })
+  }
+
+  async proposalVotes(): Promise<string[]> {
+    const revs = await computer.getTXOs({ mod: this.proposalMod })
+    const voteTxIdsSet = new Set<string>(revs.map((r) => r.split(':')[0]))
+    const validVotes = new Set<string>(voteTxIdsSet)
+
+    for (const voteTxId of voteTxIdsSet) {
+      const ancestors = (await computer.getAncestors(voteTxId)) as string[]
+      const ancestorsSet = new Set<string>(ancestors)
+      ancestorsSet.delete(voteTxId)
+
+      if ([...ancestorsSet].some(voteTxIdsSet.has, voteTxIdsSet)) {
+        validVotes.delete(voteTxId)
+      }
+    }
+
+    return Array.from(validVotes)
+  }
+
+  private async validVotes(): Promise<Vote[]> {
+    const proposalVotes = await this.proposalVotes()
+
+    const resolved = (await Promise.all(proposalVotes.map((txId) => computer.sync(txId)))).map(
+      // @ts-expect-error type unknown
+      (obj) => obj.res,
+    )
+
+    return [...resolved].filter(
+      (r: Vote) => r.electionId === this._id && r.tokenRoot === this.tokenRoot,
+    )
```

Comment:
In addition we have to check that all transactions with id in `proposalVote` have the correct expression (for example using `computer.decode` with the inner computer). Otherwise someone can just use a class that sets `electionId` and `tokenRoot` as desired and `tokenAmount` to an arbitrarily large number.

Reply:
My suggestion would be to come back to this comment once we have refactored according to the other comments.